### ISSUE TITLE
Ignore exports used in tests

### ISSUE THREAD

**Issue by alecmev** [webpro/knip#193]
I have a setup like so:

```ts
// src/foo.ts
export const bar = () => { /* ... */ }
export const foo = () => bar()
```

```ts
// src/foo.test.ts
import { foo, bar } from "./foo.js"
// Test both of them
```

```ts
// src/entry.ts
export { foo } from "./foo.js"
```

`--production --strict` reports `bar` as unused, but I do need to export it to test it. I'm aware of `ignoreExportsUsedInFile`, but I actually like being told about things that don't need to be exported. I'm also aware of `@public`, but it isn't public, and Knip has all the information necessary to deduct this without any configuration, in theory.

**Comment by webpro** [comment]
Knip would have to match internal files using the exported `bar` against the test file patterns. Sounds like a good idea.

Initially plugins like Jest and Playwright defined their own (extra) test (entry) file patterns, but later I decided to move them to make this a single item. Users can then still configure custom `entry` as needed. So if users would do this, it would complicate matters of the general idea here (there's no way to distinguish `entry` from test files in that case), although it's probably a bit of an edge case.

**Comment by alecmev** [comment]
I agree with that in principle, but I think that should be enforced separately, by some sort of "don't test internal APIs" check. Which should be possible to disable with a dedicated pragma or something like that, as there are always exceptions. For example, in our case we have an internal cubic polynomial root finder, and a bunch of higher-level functions making use of it, but it would be a tremendous effort to test its edge cases through those functions, compared to copy-pasting from scientific papers.

**Comment by webpro** [comment]
I think it should be opt-in, for those who really need it.

**Comment by TkDodo** [comment]
> I think it should be opt-in, for those who really need it.

sounds perfect 👍

**Comment by webpro** [comment]
> Knip would have to match internal files using the exported bar against the test file patterns. Sounds like a good idea.

Looking into it, this is actually not feasible. The (imports of the) test files are not even available (i.e. not part of the `ts.program()`), so there's no way to know in this particular case whether `bar` is used or not (without going crazy and defeating the purpose of `--production`).

Knip can look at the `@internal` tag, but the only way I could see this working is like this:

- In normal mode, the `@internal` tag is ignored, normal checks apply (i.e. no changes)
- In `--production` mode, exports tagged `@internal` are always ignored (and thus not reported)

Which makes it maybe usable for the case as presented, but has the obvious downside that unused `@internal` exports may be forgotten about. It feels half-baked.

So at this point I feel like closing this with a "wontfix", but happy to hear any thoughts to convince me otherwise.

(By the way, the `--strict` flag is not relevant here.)

**Comment by alecmev** [comment]
> The (imports of the) test files are not even available (i.e. not part of the `ts.program()`) ...

Hmm. Is it because the test files aren't marked as entry points? Could they be?

> ... unused `@internal` exports may be forgotten about.

Agreed. I have a similar feeling about `@public`.

**Comment by webpro** [comment]
> Hmm. Is it because the test files aren't marked as entry points? Could they be?

The idea of `--production` mode is to leave out test files. So, no.

It jokingly crossed my mind to add `--production --loose`. It's not serious, but I do think the use case is valid. I even added `@public` to the Knip codebase for the same reason: https://github.com/webpro/knip/blob/ff0710a5abe68a1bc9fe00907f714f8b8a36ae40/src/plugins/eslint/helpers.ts#L82-L86

In this particular case I could (and I will) easily resolve it by moving the test cases to the fixtures, etc. but I think there are valid cases where exporting internal things for the sake of e.g. testability is just pragmatic and useful.

I think the `@internal` tag is a good way to express this. Maybe Knip needs another mode to handle this, but might make it even harder for users to grasp all those flags. Will give this some more thought and investigate when I have more time. Feedback/ideas always welcome :)

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v2.20.0. See [Release 2.20.0](https://github.com/webpro/knip/releases/tag/2.20.0) for release notes._
